### PR TITLE
Added GarbageCollector that runs garbage collection on the GC nodes generated during summarize

### LINKED ISSUE #4056: Add / run main garbage collection logic
Implement the main garbage collection logic. The code to run it can be added either with this change or can be broken into a separate change if this one becomes too big.

### PR DESCRIPTION
**Pull request by agarwal-navin**
- Added `GarbageCollector` class to runtime utils. It can run garbage collection on a given reference graph.
- ContainerRuntime creates a garbage collector during its creation. It calls into it to run GC during summarize with the GC nodes in the summarize result.
- Added tests for GarbageCollector.

Fixes #4056

### CODE REVIEW

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 1]
```diff
@@ -1359,10 +1355,14 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
             }
 
             const trace = Trace.start();
-            const treeWithStats = await this.summarize(fullTree || safe, true /* trackState */);
+            const summarizeResult = await this.summarize(fullTree || safe, true /* trackState */);
+
+            // Run garbage collection on the GC nodes returned by summarize.
+            const startingRoutes = [ "/" ];
+            this.garbageCollector.runGC(summarizeResult.gcNodes, startingRoutes);
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1362]: After GC has run, data stores should be notified of routes that are used in it. I will address that separately. Created a work item to track this - #4493

File: packages/runtime/container-runtime/src/dataStoreContext.ts [hunk 2]
```diff
@@ -179,6 +179,11 @@ export abstract class FluidDataStoreContext extends TypedEventEmitter<IFluidData
         return this.registry;
     }
 
+    public async getIsRoot(): Promise<boolean> {
+        const initialDetails = await this.getInitialSnapshotDetails();
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/dataStoreContext.ts:183]: This will be updated to use the data from "GC" blob once #4484 is checked in.

File: packages/runtime/runtime-utils/src/index.ts [hunk 3]
```diff
@@ -6,6 +6,7 @@
 export * from "./dataStoreHandleContextUtils";
 export * from "./dataStoreHelpers";
 export * from "./garbageCollectionUtils";
+export * from "./garbageCollector";
```
Comment (anthony-murphy) [inline packages/runtime/runtime-utils/src/index.ts:9]: this doesn't really seem like a util to me. actually most of these don't see like utils. to me utils should be things used by the container runtime, but also might be used independently, like by data stores.

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 4]
```diff
@@ -915,13 +921,13 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
         return root;
     }
 
-    protected serializeContainerBlobs(summaryTreeBuilder: SummaryTreeBuilder) {
+    protected addContainerBlobsToSummary(summaryTree: ISummaryTreeWithStats) {
```
Comment (anthony-murphy) [inline packages/runtime/container-runtime/src/containerRuntime.ts:924]: >protected [](start = 4, length = 9)

why is this protected?

**Review by anthony-murphy — APPROVED** [review]
:shipit:

File: packages/runtime/runtime-utils/src/index.ts (hunk 3 above)
Comment (agarwal-navin) [reply packages/runtime/runtime-utils/src/index.ts:9]: The garbageCollector is independent of container runtime or any other runtime concepts. The idea I had in mind is that any one who wishes to run GC on its data can use this garbage collector
 For example, a collection data store can use it to run GC to delete unused instances. Any data store or blob manager can also use it.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 4 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts:924]: Not sure. This and `getDataStore` are the only protected members. Doesn't look like its for inheritance.
@vladsud added both these methods so he can probably help here.

File: packages/runtime/runtime-utils/src/index.ts (hunk 3 above)
Comment (anthony-murphy) [reply packages/runtime/runtime-utils/src/index.ts:9]: should it just be its own package then? i just don't want utils to be a dumping ground which is what frequently happens with utils packages

Comment (anthony-murphy) [reply packages/runtime/runtime-utils/src/index.ts:9]: no need to block on it. just something to consider

File: packages/runtime/runtime-utils/src/garbageCollector.ts [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+export interface IGarbageCollector {
+    /**
+     * Runs GC on the given reference graph. startingIds are ids of nodes from which the GC algorithm starts
+     * and searches for other referenced nodes.
+     */
+    runGC(referenceGraph: IGraphNode[], startingIds: string[]): IGCResult;
+}
+
+export class GarbageCollector implements IGarbageCollector {
```
Comment (anthony-murphy) [inline packages/runtime/runtime-utils/src/garbageCollector.ts:22]: seems like this could just be a function, rather than a class. you coulds still have the interface/type that defines the signature if that's needed

**Review by anthony-murphy — APPROVED** [review]

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 4 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/containerRuntime.ts:924]: I do not remember adding this method :) Nor I'm sure if the question about them being better private or public?

File: packages/runtime/runtime-utils/src/garbageCollector.ts [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    referencedNodes: IGraphNode[];
+    deletedNodes: IGraphNode[];
+}
+
+export interface IGarbageCollector {
+    /**
+     * Runs GC on the given reference graph. startingIds are ids of nodes from which the GC algorithm starts
+     * and searches for other referenced nodes.
+     */
+    runGC(referenceGraph: IGraphNode[], startingIds: string[]): IGCResult;
```
Comment (vladsud) [inline packages/runtime/runtime-utils/src/garbageCollector.ts:19]: should startingIds be rootIds? It's not very clear what "starting" means in this context

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 7]
```diff
@@ -1359,10 +1355,14 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
             }
 
             const trace = Trace.start();
-            const treeWithStats = await this.summarize(fullTree || safe, true /* trackState */);
+            const summarizeResult = await this.summarize(fullTree || safe, true /* trackState */);
+
+            // Run garbage collection on the GC nodes returned by summarize.
+            const startingRoutes = [ "/" ];
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1361]: can we get rid of startingRoutes variable and just pass ["/"] to a function?

File: packages/runtime/container-runtime/src/dataStoreContext.ts [hunk 8]
```diff
@@ -179,6 +179,11 @@ export abstract class FluidDataStoreContext extends TypedEventEmitter<IFluidData
         return this.registry;
     }
 
+    public async getIsRoot(): Promise<boolean> {
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/dataStoreContext.ts:182]: I'd call it isRoot(). I do not thing "get" adds anything here.

File: packages/runtime/container-runtime/src/dataStores.ts [hunk 9]
```diff
@@ -435,19 +443,23 @@ export class DataStores implements IDisposable {
...
 
     /**
      * @returns this channel's garbage collection node.
      */
-    private getGCNode(): IGraphNode {
+    private async getGCNode(): Promise<IGraphNode> {
         /**
-         * Get the outbound routes of this channel. This will be updated to only consider root data stores
-         * as referenced and hence outbound.
+         * Get the outbound routes of this channel. Only root data stores are considered referenced.
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/dataStores.ts:455]: This comment should be above function.

File: packages/runtime/container-runtime/src/dataStores.ts [hunk 10]
```diff
@@ -377,11 +379,11 @@ export class DataStores implements IDisposable {
...
     }
 
-    public async summarizeInternal(
-        builder: SummaryTreeBuilder, fullTree: boolean, trackState: boolean): Promise<void> {
+    public async summarize(fullTree: boolean, trackState: boolean): Promise<IChannelSummarizeResult> {
+        const builder = new SummaryTreeBuilder();
         // A list of this channel's GC nodes. Starts with this channel's GC node and adds the GC nodes all its child
         // channel contexts.
-        let gcNodes: IGraphNode[] = [ this.getGCNode() ];
+        let gcNodes: IGraphNode[] = [ await this.getGCNode() ];
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/dataStores.ts:386]: I worry that most of the code around GC assumes (without asserting) that document can't change. I.e. ops are paused, no external users can add / delete anything (data store / dds).
It would be great to sprinkle code with asserts in places where such assumption is the core for correctness. I worry if we need to change design (let's say in future get rid of summarizer client and teach main client to run GC in presence of local changes, without pausing ops), that it would be extremely hard to impossible to extract all assumptions in GC-related code and not to break them
Basically, defense in depth, fail fast.

File: packages/runtime/runtime-utils/src/garbageCollector.ts (hunk 5 above)
Comment (vladsud) [reply packages/runtime/runtime-utils/src/garbageCollector.ts:22]: +1. Class assumes state. I think the design is for GC to be stateless (between summaries)

File: packages/runtime/runtime-utils/src/garbageCollector.ts [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+     * Runs GC on the given reference graph.
+     * @param referenceGraph - The reference graph to run GC on.
+     * @param startingIds - The ids of initial nodes where GC starts its run.
+     * @returns a list of referenced nodes and a list of deleted nodes in the reference graph.
+     */
+    public runGC(referenceGraph: IGraphNode[], startingIds: string[]): IGCResult {
+        // Create a map of node id to node from the given reference Graph for quick node lookup.
+        const referenceMap = this.createReferenceMap(referenceGraph);
+        // This map keeps track of nodes that we have visited. It is used to detect cycles in the graph.
+        const visited: Map<string, boolean> = new Map();
```
Comment (vladsud) [inline packages/runtime/runtime-utils/src/garbageCollector.ts:35]: Please use set, no need for map

**Review by vladsud — APPROVED** [review]

File: packages/runtime/runtime-utils/src/index.ts (hunk 3 above)
Comment (agarwal-navin) [reply packages/runtime/runtime-utils/src/index.ts]: I moved it to a new package under packages/utils. Please let me know if this looks good.

File: packages/runtime/runtime-utils/src/garbageCollector.ts (hunk 5 above)
Comment (agarwal-navin) [reply packages/runtime/runtime-utils/src/garbageCollector.ts]: Changed it to a function.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 4 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts]: I changed it to private.

File: packages/runtime/container-runtime/src/dataStores.ts (hunk 10 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/dataStores.ts:386]: I agree. However today, summarize can be called from multiple places and I am not sure if all of them pause delta manager, etc. I believe this will need some thorough investigation before we can add such fail fasts.

I will add such asserts for GC in one of my subsequent PRs where I separate out GC and summarize.

File: packages/runtime/runtime-utils/src/index.ts (hunk 3 above)
Comment (agarwal-navin) [reply packages/runtime/runtime-utils/src/index.ts]: I had to move IGraphNode to the new utils as well because of layer violations and cycling dependencies.

**Comment by agarwal-navin** [comment]
Added a flag to `IContainerRuntimeOptions` that can be used to disable running GC.

File: packages/runtime/garbage-collector/src/garbageCollector.ts [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+        }
+    }
+
+    return { referencedNodes, deletedNodes };
+}
+
+function createReferenceMap(referenceGraph: IGraphNode[]) {
+    const referencedMap: Map<string, IGraphNode> = new Map();
+    for (const node of referenceGraph) {
+        referencedMap.set(node.id, node);
```
Comment (anthony-murphy) [inline packages/runtime/garbage-collector/src/garbageCollector.ts:69]: as i commented in other PR this could just be a map, then you wouldn't need to convert it to one

Comment (agarwal-navin) [reply packages/runtime/garbage-collector/src/garbageCollector.ts:69]: good point. I will work on this in a subsequent PR once I get this in.
